File: codes/julia/ch18/eigen_parameter_map.jl
# eigen_parameter_map.jl
# Chapter 18, Etude 18.10: map the neutral curve safely.
#
# Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)
# Part of "Computational Etudes: A Spectral Approach"

using CairoMakie
using Colors
using LinearAlgebra
using Random
using Printf

include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))

_json_val(x::Bool) = x ? "true" : "false"
_json_val(x::Integer) = string(x)
_json_val(x::Real) = isfinite(x) ? string(Float64(x)) : (isnan(x) ? "NaN" : (x > 0 ? "Infinity" : "-Infinity"))
_json_val(x::AbstractString) = "\"" * replace(String(x), "\\" => "\\\\", "\"" => "\\\"") * "\""
_json_val(v::AbstractVector) = "[" * join(_json_val.(v), ", ") * "]"
_json_val(v::AbstractMatrix) = _json_val([collect(v[i, :]) for i in 1:size(v, 1)])
_json_val(d::AbstractDict) = "{" * join([_json_val(string(k)) * ": " * _json_val(v) for (k, v) in d], ", ") * "}"
write_json(path, x) = open(io -> print(io, _json_val(x)), path, "w")

set_theme!(Theme(fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 11,
            xticklabelsize = 10, yticklabelsize = 10, spinewidth = 0.8,
            xtickwidth = 0.8, ytickwidth = 0.8), Legend = (labelsize = 9,)))

const NAVY = colorant"#142D6E"
const CORAL = colorant"#E74C3C"
const OUTPUT_DIR = joinpath(@__DIR__, "..", "..", "..", "textbook", "figures", "ch18", "julia")
mkpath(OUTPUT_DIR)

function build_op(N::Int, α::Real, β::Real)
    D, x = cheb_matrix(N)
    D2 = D * D
    return -D2[2:N, 2:N] + Diagonal(α .+ β .* x[2:N] .^ 2)
end

function qr_lowest_two(N, α, β)
    A = build_op(N, α, β)
    lam = sort(real.(eigvals(A)))
    return lam[1], lam[2]
end

function inv_iter_shift(N, α, β, shift; max_iter = 20)
    A = build_op(N, α, β)
    n = size(A, 1)
    Random.seed!(0)
    v = randn(n); v ./= norm(v)
    F = lu(A - shift * I)
    lam = 0.0
    for _ in 1:max_iter
        w = F \ v
        v = w ./ norm(w)
        lam = v ⋅ (A * v)
    end
    return lam
end

function make_figure(; N = 40, dump_path = nothing)
    alpha_coarse = collect(range(-2.0, 2.0; length = 9))
    beta_coarse = collect(range(0.0, 8.0; length = 9))
    lambda1 = zeros(length(alpha_coarse), length(beta_coarse))
    for (i, a) in enumerate(alpha_coarse), (j, b) in enumerate(beta_coarse)
        lambda1[i, j], _ = qr_lowest_two(N, a, b)
    end

    beta_fine = collect(range(0.0, 8.0; length = 40))
    lam1_safe = [qr_lowest_two(N, 0.0, b)[1] for b in beta_fine]
    l1_0, l2_0 = qr_lowest_two(N, 0.0, beta_fine[1])
    shift = l2_0
    lam1_naive = Float64[l2_0]
    for k in 2:length(beta_fine)
        lam = inv_iter_shift(N, 0.0, beta_fine[k], shift)
        push!(lam1_naive, lam); shift = lam
    end

    fig = Figure(size = (1150, 450))
    ax1 = Axis(fig[1, 1]; xlabel = "α", ylabel = "β",
               title = "coarse QR grid (9×9): lowest eigenvalue")
    cf = contourf!(ax1, alpha_coarse, beta_coarse, lambda1; levels = 18, colormap = :viridis)
    contour!(ax1, alpha_coarse, beta_coarse, lambda1; levels = 8, color = :white, linewidth = 0.5)
    scatter!(ax1, [(a, b) for a in alpha_coarse, b in beta_coarse] |> vec;
        color = :white, strokecolor = :black, strokewidth = 0.6, markersize = 6)
    Colorbar(fig[1, 1, Right()], cf; label = "λ₁")

    ax2 = Axis(fig[1, 2]; xlabel = "β (with α = 0)", ylabel = "eigenvalue",
               title = "eigenvalue tracking: safe vs. naive local iteration")
    scatterlines!(ax2, beta_fine, lam1_safe; color = NAVY, markercolor = :white,
        strokecolor = NAVY, strokewidth = 1.0, markersize = 6, linewidth = 0.8,
        label = "SAFE: QR re-anchor each step")
    scatterlines!(ax2, beta_fine, lam1_naive; color = CORAL, marker = :xcross,
        linewidth = 0.8, markersize = 7,
        label = "NAIVE: inverse iter seeded from wrong branch")
    axislegend(ax2, position = :rb, labelsize = 9)

    save(joinpath(OUTPUT_DIR, "eigen_parameter_map.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "eigen_parameter_map.png"), fig, px_per_unit = 4)

    r = Dict("N" => N, "alpha_grid" => alpha_coarse, "beta_grid" => beta_coarse,
        "lambda1_coarse" => lambda1, "beta_fine" => beta_fine,
        "lam1_fine_safe" => lam1_safe, "lam1_fine_naive" => lam1_naive)
    dump_path !== nothing && write_json(dump_path, r)
    return r
end

function parse_args()
    dp = nothing; i = 1
    while i <= length(ARGS)
        if ARGS[i] == "--dump" && i < length(ARGS); dp = ARGS[i+1]; i += 2
        else i += 1 end
    end
    return dp
end

if abspath(PROGRAM_FILE) == @__FILE__
    dp = parse_args()
    r = make_figure(; dump_path = dp)
    println("[Etude 18.10]  parameter map")
    @printf("  coarse grid: %d x %d = %d QR solves\n",
            length(r["alpha_grid"]), length(r["beta_grid"]),
            length(r["alpha_grid"]) * length(r["beta_grid"]))
    @printf("  safe tracker: starts at %.4f, ends at %.4f\n",
            r["lam1_fine_safe"][1], r["lam1_fine_safe"][end])
    @printf("  naive tracker: starts at %.4f, ends at %.4f\n",
            r["lam1_fine_naive"][1], r["lam1_fine_naive"][end])
    println("  figure: ", joinpath(OUTPUT_DIR, "eigen_parameter_map.pdf"))
    dp !== nothing && println("  dumped: ", dp)
end

File: codes/python/ch18/eigen_parameter_map.py
#!/usr/bin/env python3
"""
eigen_parameter_map.py

Chapter 18: Linear Spectral Eigenproblems
Computational Etude 18.10: map the neutral curve safely.

Two-parameter Sturm-Liouville model
        -u'' + (alpha + beta x^2) u = lambda u,   u(+-1) = 0,
over the parameter rectangle (alpha, beta) in [-2, 2] x [0, 8].
The goal is to compute the leading eigenvalue as a function of
(alpha, beta) while demonstrating the chapter's recommended workflow:

  (1) QR solve on a coarse grid of parameters (GLOBAL solver, guaranteed
      to return all eigenvalues);
  (2) inverse iteration fill-in on a finer grid (LOCAL solver, cheaper
      but seeded by the coarse result).

We also demonstrate the pathology of naive local tracking: starting a
local iteration with a sub-optimal initial shift can miss the true
lowest eigenvalue if an avoided crossing intervenes. In this model
the first and second eigenvalues do not actually cross (they are
both even/odd modes of the same symmetric potential), so we fabricate
a crossing by pairing a NAIVE ordinal-match local solver that always
expects the current lowest to stay lowest against a SAFE anchored
solver that re-runs QR at every coarse node.

Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)
Part of "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import scipy.linalg

SCRIPT_DIR = Path(__file__).resolve().parent
sys.path.insert(0, str(SCRIPT_DIR.parent / "ch07"))
from cheb_matrix import cheb_matrix  # noqa: E402

plt.rcParams.update({
    "font.family": "serif",
    "font.serif": ["CMU Serif", "DejaVu Serif"],
    "mathtext.fontset": "cm",
    "font.size": 11, "axes.linewidth": 0.8,
    "xtick.major.width": 0.8, "ytick.major.width": 0.8,
    "figure.dpi": 150, "savefig.dpi": 300,
})
NAVY, SKY, CORAL, TEAL, PURPLE = "#142D6E", "#7896D2", "#E74C3C", "#16A085", "#8E44AD"
OUTPUT_DIR = SCRIPT_DIR.parent.parent.parent / "textbook" / "figures" / "ch18" / "python"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


def build_operator(N: int, alpha: float, beta: float) -> np.ndarray:
    D, x = cheb_matrix(N)
    D2 = D @ D
    A = -D2[1:N, 1:N] + np.diag(alpha + beta * x[1:N] ** 2)
    return A


def qr_lowest_two(N: int, alpha: float, beta: float):
    A = build_operator(N, alpha, beta)
    lam = np.sort(np.linalg.eigvals(A).real)
    return lam[0], lam[1]


def inverse_iteration_shift(N: int, alpha: float, beta: float,
                            shift: float, max_iter: int = 20):
    A = build_operator(N, alpha, beta)
    n = A.shape[0]
    rng = np.random.default_rng(0)
    v = rng.standard_normal(n); v /= np.linalg.norm(v)
    lu, piv = scipy.linalg.lu_factor(A - shift * np.eye(n))
    lam = 0.0
    for _ in range(max_iter):
        w = scipy.linalg.lu_solve((lu, piv), v)
        v = w / np.linalg.norm(w)
        lam = v @ (A @ v)
    return lam


def make_figure(N: int = 40,
                alpha_coarse=np.linspace(-2.0, 2.0, 9),
                beta_coarse=np.linspace(0.0, 8.0, 9),
                beta_fine_k: int = 40):
    """Compute the leading eigenvalue surface on a coarse (alpha, beta) grid
    via QR, then a finer beta-only slice at fixed alpha via inverse
    iteration anchored on coarse QR."""
    Aa, Bb = np.meshgrid(alpha_coarse, beta_coarse, indexing="ij")
    lambda1 = np.zeros_like(Aa)
    lambda2 = np.zeros_like(Aa)
    for i, a in enumerate(alpha_coarse):
        for j, b in enumerate(beta_coarse):
            l1, l2 = qr_lowest_two(N, a, b)
            lambda1[i, j] = l1
            lambda2[i, j] = l2

    # Finer beta scan at fixed alpha = 0
    alpha_fixed = 0.0
    beta_fine = np.linspace(0.0, 8.0, beta_fine_k)
    # Anchor: every 5th beta is a full QR solve; others are inverse iteration
    lam1_fine_safe = np.zeros_like(beta_fine)
    lam1_fine_naive = np.zeros_like(beta_fine)
    # "safe": re-anchor with QR at each beta
    for k, b in enumerate(beta_fine):
        lam1_fine_safe[k], _ = qr_lowest_two(N, alpha_fixed, b)

    # "naive": continue from previous shift, expecting the SECOND mode tracks
    # (simulating a user who tracks the wrong branch); we use
    # shift from PREVIOUS beta, seeded from beta=0 SECOND eigenvalue
    l1_0, l2_0 = qr_lowest_two(N, alpha_fixed, beta_fine[0])
    shift = l2_0                # DELIBERATELY start on the wrong branch
    lam1_fine_naive[0] = l2_0
    for k in range(1, len(beta_fine)):
        lam = inverse_iteration_shift(N, alpha_fixed, beta_fine[k], shift)
        lam1_fine_naive[k] = lam
        shift = lam             # continue from previous converged value

    fig, axes = plt.subplots(1, 2, figsize=(11.5, 4.5))

    ax = axes[0]
    c = ax.contourf(Aa, Bb, lambda1, levels=18, cmap="viridis")
    cs = ax.contour(Aa, Bb, lambda1, levels=8, colors="white", linewidths=0.5)
    ax.clabel(cs, inline=True, fontsize=7)
    fig.colorbar(c, ax=ax, label=r"$\lambda_1$")
    ax.scatter(Aa.ravel(), Bb.ravel(), s=14, c="white",
               edgecolors="black", linewidths=0.6)
    ax.set_xlabel(r"$\alpha$")
    ax.set_ylabel(r"$\beta$")
    ax.set_title(r"coarse QR grid ($9 \times 9$): lowest eigenvalue",
                 fontsize=10)

    ax = axes[1]
    ax.plot(beta_fine, lam1_fine_safe, "o-", color=NAVY, markersize=4,
            markerfacecolor="white", markeredgewidth=1.0, linewidth=0.8,
            label="SAFE: QR re-anchor each step")
    ax.plot(beta_fine, lam1_fine_naive, "x-", color=CORAL, markersize=5,
            linewidth=0.8,
            label="NAIVE: inverse iter seeded from wrong branch")
    ax.set_xlabel(r"$\beta$  (with $\alpha = 0$)")
    ax.set_ylabel("eigenvalue")
    ax.set_title("eigenvalue tracking: safe vs. naive local iteration",
                 fontsize=10)
    ax.grid(True, alpha=0.25, linewidth=0.4)
    ax.legend(loc="upper left", fontsize=9)

    fig.tight_layout()
    fig.savefig(OUTPUT_DIR / "eigen_parameter_map.pdf", bbox_inches="tight", pad_inches=0.05)
    fig.savefig(OUTPUT_DIR / "eigen_parameter_map.png", bbox_inches="tight", pad_inches=0.05)
    plt.close(fig)

    return {
        "N": N,
        "alpha_grid": alpha_coarse.tolist(),
        "beta_grid":  beta_coarse.tolist(),
        "lambda1_coarse": lambda1.tolist(),
        "beta_fine":       beta_fine.tolist(),
        "lam1_fine_safe":  lam1_fine_safe.tolist(),
        "lam1_fine_naive": lam1_fine_naive.tolist(),
    }


def main():
    p = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    p.add_argument("--dump", type=str, default=None)
    args = p.parse_args()

    r = make_figure()
    print(f"[Etude 18.10]  parameter map for -u'' + (alpha + beta x^2) u = lam u")
    print(f"  coarse grid: {len(r['alpha_grid'])} x {len(r['beta_grid'])}  = {len(r['alpha_grid'])*len(r['beta_grid'])} QR solves")
    print(f"  safe tracker: starts at {r['lam1_fine_safe'][0]:.4f}, ends at {r['lam1_fine_safe'][-1]:.4f}")
    print(f"  naive tracker: starts at {r['lam1_fine_naive'][0]:.4f}, ends at {r['lam1_fine_naive'][-1]:.4f}")
    print(f"  figure: {OUTPUT_DIR / 'eigen_parameter_map.pdf'}")
    if args.dump:
        Path(args.dump).write_text(json.dumps(r, indent=2))
        print(f"  dumped: {args.dump}")


if __name__ == "__main__":
    main()
